services/publishing/Tweek.Publishing.Service/Helpers/HooksHelper.cs

```diff
@@ -0,0 +1,153 @@
+using App.Metrics;
+using App.Metrics.Counter;
+using Tweek.Publishing.Service.Model.Hooks;
+using Newtonsoft.Json;
+
+namespace Tweek.Publishing.Helpers {
+  public class HooksHelper {
+    private static HooksHelper _instance;
+    private readonly IMetrics _metrics;
+    private readonly ILogger _logger;
+    private readonly Func<string, Task<string>> _git;
+    private readonly Regex _keysRegex;
+    private static readonly CounterOptions _hooksMetric = new CounterOptions{Context = "publishing", Name = "hooks"};
+    private static readonly MetricTags _metricsSuccess = new MetricTags("Status", "Success");
+    private static readonly MetricTags _metricsFailure = new MetricTags("Status", "Failure");
+
+    private HooksHelper(Func<string, Task<string>> gitExecutor, IMetrics metrics, ILogger logger) {
+      this._logger = logger;
+      this._git = gitExecutor;
+      this._metrics = metrics;
+      this._keysRegex = new Regex(@"(?:implementations/jpad/|manifests/)(.*)\..*", RegexOptions.Compiled);
+    }
+
+    public static void Initialize(Func<string, Task<string>> gitExecutor, IMetrics metrics, ILogger logger = null) {
+      _instance = new HooksHelper(gitExecutor, metrics, logger ?? NullLogger.Instance);
+    }
+
+    public static HooksHelper GetInstance() {
+      return _instance;
+    }
+
+    public async Task TriggerNotificationHooksForCommit(string commitId) {
+      try {
+        var keyPaths = await _GetKeyPathsFromCommit(commitId);
+        var allHooks = await _GetAllHooks(commitId);
+
+        var keyPathsByHook = _AggregateKeyPathsByHook(keyPaths, allHooks);
+        var usedKeyPaths = _GetUsedKeyPaths(keyPathsByHook);
+        var keyPathsData = await _GetKeyPathsData(usedKeyPaths, commitId);
+
+        var hooksWithData = _GetHooksWithKeyPathData(keyPathsByHook, keyPathsData);
+        await _TriggerHooks(hooksWithData, commitId);
+      } catch (Exception ex) {
+        _logger.LogError(ex, $"Failed triggering notification hooks for commit {commitId}");
+        _metrics.Measure.Counter.Increment(_hooksMetric, _metricsFailure);
+      }
+    }
+
+    private async Task _TriggerHooks(Dictionary<Hook, string> hooksWithData, string commitId) {
+      var triggerTasks = hooksWithData.Select( kvp => kvp.Key.Trigger(kvp.Value) );
+
+      foreach (var triggerTask in triggerTasks) {
+        try {
+          await triggerTask;
+
+          _metrics.Measure.Counter.Increment(_hooksMetric, _metricsSuccess);
+        } catch (Exception ex) {
+          _logger.LogError(ex, $"Failed triggering a notification hook for commit {commitId}");
+          _metrics.Measure.Counter.Increment(_hooksMetric, _metricsFailure);
+        }
+      }
+    }
+
+    private Dictionary<Hook, string> _GetHooksWithKeyPathData(
+      Dictionary< Hook, HashSet<string> > keyPathsByHook,
+      Dictionary<string, KeyPathData> keyPathsData
+    ) {
+      return keyPathsByHook.Aggregate(new Dictionary<Hook, string>(), (acc, kvp) => {
```

Comment:
Can be re-written with ToDictionary(). Might be simpler and remove the ".Add" part